### ISSUE TITLE
Calibrate order tree critical height

### ISSUE THREAD

**Issue by alnoki** [econia-labs/econia#27]
TL; DR 3 direct questions at the bottom

# AVL queue structure

Econia uses a custom data structure, the [AVL queue], for storing orders.
In short, the AVL queue combines an AVL tree with a doubly linked list at every tree node, where tree nodes are price levels and list nodes are orders:

>                                        1001 [35 -> 38]
>                                       /    \
>                   [50 -> 60 -> 55] 1000    1003 [20]
>     AVL queue head ^                      /    \
>                              [15 -> 5] 1002    1004 [4 -> 10]
>                                                           ^ AVL queue tail

Here, for an "ascending" AVL queue, orders are sorted by:

1. Increasing price, then
2. Increasing order of insertion

Hence a large taker order (in this case a market buy) fills against the 3 orders at price 1000 (size 50, size 60, then size 55), then against the 2 orders at price 1001 (size 35, then size 38), and so on.
Conversely, for a "descending" AVL queue, orders are sorted by:

1. Decreasing price, then
2. Increasing order of insertion

>                            1001 [35 -> 38]
>                           /    \
>       [50 -> 60 -> 55] 1000    1003 [20]
>     AVL queue tail ^          /    \
>                  [15 -> 5] 1002    1004 [4 -> 10]
>                                          ^ AVL queue head

Here, a large taker order (in this case a market sell) fills against the 2 orders at price 1004 (size 4 then size 10), then against the single order at price 1003 (size 20), and so on.
Hence a descending AVL queue constitutes a bids tree, which in this case, corresponds to the following book structure with price-time priority.

| Price | Size |
|-------|------|
| 1004  | 4    |
| 1004  | 10   |
| 1003  | 20   |
| 1002  | 15   |
| 1002  | 5    |
| 1001  | 35   |
| 1001  | 38   |
| 1000  | 50   |
| 1000  | 60   |
| 1000  | 55   |

# Insertions

When a new order is placed, it is inserted at the tail of the corresponding doubly linked list for the given price level, if such a price level is already in the tree.
For instance, continuing the above example, placing an order for size 18 at price 1003 would lead to the following, in either an ascending or descending AVL queue:

>                          1001 [35 -> 38]
>                         /    \
>     [50 -> 60 -> 55] 1000    1003 [20 -> 18]
>                             /    \       ^ new list node
>                [15 -> 5] 1002    1004 [4 -> 10]

Attempting to place an order at a price of 1005, however, would require inserting a new tree node, yielding an unbalanced AVL tree:

>         1001
>        /    \
>     1000    1003
>            /    \
>         1002    1004
>                     \
>                     1005

And this would require a rotation:

>             1003
>            /    \
>         1001    1004
>        /    \       \
>     1000    1002    1005

This self-balancing behavior reduces lookup cost reductions, since the upper bound on AVL tree height is approximately $1.44 \log_2 n - 1$ (see [AVL queue spec height section] for supporting calculations), but without a bounded number of possible price levels, the tree can still grow to prohibitive sizes.
In particular, with Aptos' storage gas schedule charging per-item costs, lookup gas costs increase linearly with tree height.

# Eviction calibration

To prevent tree height from growing too large, the [AVL queue] incorporates eviction functionality based on a critical height:
if tree height is above a critical threshold, then inserting to the AVL queue automatically evicts the order at the AVL queue tail, having the lowest price-time priority.
For instance, returning to the first example above, at a critical height of 2, inserting an order at a price of 1001 would evict the order with price 1004, size 10 in an ascending AVL queue, and in a descending AVL queue, would evict the order with price 1000, size 55.
Since multiple orders can be inserted into a single price level, this eviction paradigm only affects the number of price levels, not the number of orders on the book.

As for the optimal critical tree height, this is a trade-off between gas costs and price level comprehensiveness.
If the critical tree height is too low, there will be too few price levels.
If the critical tree height is too large, gas costs will be permitted to grow inappropriately large.
As described in the [AVL queue spec height section], there is an upper bound and lower bound on the number of price levels for a given tree height:

| Height       | Min levels   | Max levels   |
|--------------|--------------|--------------|
| 0            | 1            | 1            |
| 1            | 2            | 3            |
| 2            | 4            | 7            |
| 3            | 7            | 15           |
| 4            | 12           | 31           |
| 5            | 20           | 63           |
| 6            | 33           | 127          |
| 7            | 54           | 255          |
| 8            | 88           | 511          |
| 9            | 143          | 1023         |
| 10           | 232          | 2047         |
| 11           | 376          | 4095         |
| 12           | 609          | 8191         |
| 13           | 986          | 16383        |
| 14           | 1596         | 32767        |
| 15           | 2583         | 65535        |

Hence, for instance, if critical height is set at 8, then the tree may end up with as few as 88 price levels or as many as 511 price levels in theory, and probably somewhere in the middle in practice.
To minimize gas costs while ensuring adequate market structure, the following design questions are thus put forth:

1. How many active price levels, per side, is typical of on-chain order books?
2. What is the minimum number of active price levels, per side, required to support adequate market structure?
3. Would a critical height of 6, supporting as few at 33 price levels per side, as many as 127 price levels per side, and likely some amount in the middle during steady state, support adequate market structure?

[AVL queue]:                     https://github.com/econia-labs/econia/blob/main/src/move/econia/doc/avl_queue.md
[AVL queue spec height section]: https://github.com/econia-labs/econia/blob/main/src/move/econia/doc/avl_queue.md#height

**Comment by alnoki** [comment]
# Price level coverage

Depending on the tick size for a market, the chosen critical height results in a guaranteed price level coverage. For instance, if tick size is chosen as 5 basis points of the nominal price, and critical height is chosen as 10, then at a minimum price level coverage is guaranteed for 0.05% * 232 = 11.6% depth on a given side.

With an upper bound of 1023 price levels for a critical height of 10, this corresponds to 0.05% * 1023 = 51.15% price level coverage, but this assumes 0 price level sparsity. If price levels are sparse, then even more depth can be covered.

As for the minimum number of price levels, this number corresponds to a minimal AVL tree, which is unlikely to be seen in practice as it essentially corresponds to a Fibonacci pattern. Conversely, the maximum number of price levels is also unlikely to be seen in practice, since it corresponds to a maximally-dense AVL tree. Hence estimates as to the maximum number of price levels to be seen at a critical height, in practice, hinge on empirical data.

# Tick dynamics

As for tick size, this parameter is chosen permissionlessly by market registrants when registering a market. Lower tick sizes will result in more price granularity near the spread, but come with the risk of undefined behavior during market volatility: if a large taker order clears out the book with all price levels centered around mid, there will be no liquidity left on the book.

Hence registrants should be advised to pick tick sizes on the order of 5 basis points, with recognized market status to be conferred only on appropriately-calibrated markets.
